pkg/operator/ceph/cluster/crash/add.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2019 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crash
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "ceph-crashcollector-controller"
+	// AppName is the value to the "app" label for the ceph-crash pods
+	AppName = "rook-ceph-crashcollector"
+	// NodeNameLabel is a node name label
+	NodeNameLabel = "node_name"
+)
+
+// Add adds a new Controller based on nodedrain.ReconcileNode and registers the relevant watches and handlers
+func Add(mgr manager.Manager) error {
+	reconcileNode := &ReconcileNode{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+	reconciler := reconcile.Reconciler(reconcileNode)
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return fmt.Errorf("failed to create a new %q. %+v", controllerName, err)
+	}
+
+	// Watch for changes to the nodes
+	specChangePredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			nodeOld := e.ObjectOld.DeepCopyObject().(*corev1.Node)
+			nodeNew := e.ObjectNew.DeepCopyObject().(*corev1.Node)
+			return !reflect.DeepEqual(nodeOld.Spec, nodeNew.Spec)
+		},
+	}
+	logger.Debugf("watch for changes to the nodes")
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{}, specChangePredicate)
+	if err != nil {
+		return fmt.Errorf("failed to watch for node changes. %+v", err)
+	}
+
+	// Watch for changes to the ceph-crash deployments
+	logger.Debugf("watch for changes to the ceph-crash deployments")
+	err = c.Watch(
+		&source.Kind{Type: &appsv1.Deployment{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+				logger.Debugf("try to reconcile deployment")
```

Comment:
I'd suggest not logging this. It could make the debug logs very noisy depending on the number of any kind of deployment.

Reply:
Yeah it's really verbose on debug. Let me remove that.